proposals/0540-jsem.rst

```diff
@@ -246,6 +273,110 @@ with the semaphore: instead of systematically releasing a token to the semaphore
+
+Currently, GHC 9.8.{1,2} and GHC 9.10.1 provide jobclient implementations based
+on ``semaphore-compat-1.0.0``, which ``cabal-install-3.12.1.0`` also uses to provide
+a jobserver implementation. ``semaphore-compat-1.0.0`` uses the problematic POSIX
+system semaphores and is susceptible to the restrictions regarding system C library
+usage by jobservers and jobclients.
+
+Since the underlying semaphore implementation is being fundamentally changed, the
+future version of ``semaphore-compat`` – presumably shipped with a future version
+of GHC and ``cabal-install`` – would be incompatible with existing jobservers and
+jobclients linking against ``semaphore-compat-1.0.0``.
+
+For instance, this means
+
+- ``cabal-install-3.12.1.0`` would be unable to compile code with the
+  ``--semaphore`` flag enabled for future versions of GHC shipping with the
+  updated ``semaphore-compat`` release. There could be a mechanism
+  in the new version of ``semaphore-compat`` (linked to by the future version of
+  GHC) to detect such a scenario and output an informative error message.
+
+- Similarly, future versions of ``cabal-install`` would be unable to use
+  the semaphore mechanism with existing versions of GHC such as 9.8.{1,2} and
+  9.10.1. However, this could be detected by ``cabal-install`` and the feature
+  automatically disabled, much like it currently is if attempted to be enabled
+  for versions of GHC that don't support ``-jsem`` such as 9.6.
+
+Backwards and forwards compatibility for the abstract semaphores provided by ``semaphore-compat``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As the previous section illustrates, breaking changes to the communication
+protocol used to implement these semaphores are painful and lead to the breakage
+of existing compilers and build tools.
+
+As such, we must strive to avoid such breaking changes as much as possible.
+In the (hopefully rare) case when such a change cannot be avoided, the change
+must meet a high bar of justification.
+
+Concretely, the communication protocol used in each major version series
+of ``semaphore-compat`` must be compatible with other versions of the
+library belonging to the same major version series. For instance, a
+``semaphore-compat-2.0.0`` must be compatible with ``semaphore-compat-2.9.9``, but
+may not be compatible with ``semaphore-compat-3.0.0``.
+
+Breaking changes to the protocol require a major version bump and need to be
+justified to and accepted by the GHC Proposals Committee.
+
+To allow breaking changes to the protocol to be detected we require that,
+starting with ``semaphore-compat-2.0.0``, the **semaphore identifier** must
+begin a a prefix ``v<N>-`` where the ``<N>`` specifies the ``semaphore-compat``
+major version that this semaphore was created using. For instance,
+a semaphore created with ``semaphore-compat-2.0.0`` could be called
+``v2-some-semaphore12345``.
```

Comment:
Redundant with the specification? Generally I could live with a lot less text in this section, because the compat story is clear from the specification now.